### PR TITLE
[BIOMAGE-1511] - Fix select data

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-1511

#### Link to staging deployment URL 

https://ui-agi-ui540.scp-staging.biomage.net/

#### Anything else the reviewers should know about the changes here

- Fixed issue where Frequency plot does not show result for metadata
- Fixed Categorical Embedding in Plots and Tables not showing groupBy properly
- Done some refactoring of the plots to make it easier to understand

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Code updates
Have best practices and ongoing refactors being observed in this PR

- [x] Migrated any selector / reducer used to the new format
- [x] Validated that current unit tests work as expected and are enough

### Testing
- [x] Unit tests written
- [x] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [x] Deployed to staging
- [ ] Passed integration tests: https://github.com/biomage-ltd/testing/actions/runs/1398300441

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/components/plots/styling/SelectData.jsx [hunk 1]
```diff
@@ -99,10 +103,11 @@ const SelectData = (props) => {
         </Select>
       </Form.Item>
       <div>
-        Select the Cell sets or Metadata to be shown as points:
+        Select the Cell sets or Metadata to be shown as data:
       </div>
       <Form.Item>
         <Select
+          aria-label='selectPoints'
```
Comment (cosa65) [inline src/components/plots/styling/SelectData.jsx:112]: This is to identify the select, right? Is this different/better than using a `data-test-id`?

File: src/components/plots/styling/frequency/SelectCellSets.jsx [hunk 2]
```diff
@@ -5,23 +5,28 @@ import {
...
-    onUpdate, config, optionsMetadata, optionsCellSets,
+    onUpdate, config,
   } = props;
 
-  const changeClusters = (option) => {
-    const newValue = option.value.toLowerCase();
-    onUpdate({ proportionGrouping: newValue });
-  };
+  const optionsMetadata = useSelector(getCellSetsHierarchyByType(['metadataCategorical']));
+  const optionsCellSets = useSelector(getCellSetsHierarchyByType(['cellSets']));
```
Comment (cosa65) [inline src/components/plots/styling/frequency/SelectCellSets.jsx]: Nicee

Comment (cosa65) [reply src/components/plots/styling/frequency/SelectCellSets.jsx]: I am wondering if perhaps we should make `byType` work with just receiving a single a key instead of an array, since we generally either want to get one of the two types or all (two) types (Which is `getCellSetsHierarchy`)

Comment (cosa65) [reply src/components/plots/styling/frequency/SelectCellSets.jsx]: And then rename `ByKey` to `ByKeys` to differentiate these since that still has use cases for array I think

File: src/components/plots/styling/frequency/SelectCellSets.jsx [hunk 3]
```diff
@@ -60,15 +64,15 @@ const SelectCellSets = (props) => {
         </Tooltip>
       </Form.Item>
       <div>
-        Select the cell sets to be shown:
+        Select how the data should be grouped:
```
Comment (cosa65) [inline src/components/plots/styling/frequency/SelectCellSets.jsx:67]: What is the reason for this change? was it discussed with vicky?

File: src/components/plots/styling/violin/ViolinControls.jsx [hunk 4]
```diff
@@ -22,13 +22,11 @@ const ViolinControls = (props) => {
         />
       </Panel>
       <Panel header='Select data' key='15'>
-        {config && !cellSets.loading && !cellSets.error ? (
```
Comment (cosa65) [inline src/components/plots/styling/violin/ViolinControls.jsx:25]: Removing these checks doesn't break anything?

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-categorical/index.jsx [hunk 5]
```diff
@@ -113,13 +113,11 @@ const EmbeddingCategoricalIndex = ({ experimentId }) => {
   const renderExtraPanels = () => (
     <>
       <Panel header='Select data' key='15'>
-        {config && !cellSets.loading && !cellSets.error ? (
```
Comment (cosa65) [inline src/pages/experiments/[experimentId]/plots-and-tables/embedding-categorical/index.jsx:116]: Removing these checks doesn't break anything?

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx [hunk 6]
```diff
@@ -187,6 +184,7 @@ const EmbeddingContinuousIndex = ({ experimentId }) => {
               config={config}
               onUpdate={updatePlotWithChanges}
               renderExtraPanels={renderExtraPanels}
+              defaultActiveKey={['gene-selection']}
```
Comment (cosa65) [inline src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx]: Why is `defaultActiveKey` an array only? It seems that we are always passing either a single string or a single number.

Also the `defaultActiveKey` in `Collapse` (that is the reason for this custom prop) allows for single number or single string too, I would change this so that we match the allowed types for this prop in `Collapse`. Then we can pass the single value freely in every place we are passing something different.

File: src/redux/reducers/componentConfig/initialState.js [hunk 7]
```diff
@@ -156,7 +156,6 @@ const volcanoInitialConfig = {
 const frequencyInitialConfig = {
   spec: '1.0.0',
   frequencyType: 'proportional',
-  proportionGrouping: '',
```
Comment (cosa65) [inline src/redux/reducers/componentConfig/initialState.js:159]: This was changed because we changed the name of the key for this?

File: src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js [hunk 8]
```diff
@@ -180,42 +182,67 @@ const generateSpec = (config, plotData, cellSetNames) => {
...
 
-const filterCells = (cellSets, selectedCellSet) => {
-  const newCellSets = cellSets.hierarchy.find(
-    (rootNode) => rootNode.key === selectedCellSet,
+const filterCells = (cellSets, sampleKey, groupBy) => {
+  let filteredCells = [];
+
+  // Get all the filtered cells
+  if (sampleKey === 'All') {
+    filteredCells = getAllCells(cellSets, groupBy);
```
Comment (cosa65) [inline src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:190]: Since we are refactoring things here, do you think there is a way we could refactor this so that we reuse the selectors and then pass the values to the `generateSpec` here? It will probably mean less repeated code and more memoized results usage which are good things

File: src/redux/reducers/componentConfig/initialState.js (hunk 7 above)
Comment (aerlaut) [reply src/redux/reducers/componentConfig/initialState.js:159]: this was just moved closer to `xAxisGrouping` because these 2 things are variables for similar values : how to group the data

File: src/components/plots/styling/violin/ViolinControls.jsx (hunk 4 above)
Comment (aerlaut) [reply src/components/plots/styling/violin/ViolinControls.jsx:25]: No, it's handled inside the component.

File: src/components/plots/styling/SelectData.jsx (hunk 1 above)
Comment (aerlaut) [reply src/components/plots/styling/SelectData.jsx:112]: yes. It's different from `data-test-id` because it can be read by screen readers, so it has meaning too.

File: src/components/plots/styling/frequency/SelectCellSets.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/plots/styling/frequency/SelectCellSets.jsx]: Yea I think it would be good to do that

File: src/components/plots/styling/frequency/SelectCellSets.jsx (hunk 3 above)
Comment (aerlaut) [reply src/components/plots/styling/frequency/SelectCellSets.jsx:67]: No I haven't talked with Oliver or Vicky. I think it's clearer though. I'll ask them.

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-categorical/index.jsx (hunk 5 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/embedding-categorical/index.jsx:116]: Nope, it's handled in the component.

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx (hunk 6 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx]: Yea, it's able to take either a single string or an array. I'll change the proptype checks accordingly.

File: src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js (hunk 8 above)
Comment (aerlaut) [reply src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:190]: Hmm... yea, I think I can move these 2 functions and make them into selectors.

Comment (aerlaut) [reply src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:190]: But these functions are inside `.js` files, not `.jsx` files, so we can't access selectors. I think it's cleaner to do it in another PR to refactor plots.

File: src/components/plots/styling/frequency/SelectCellSets.jsx (hunk 3 above)
Comment (aerlaut) [reply src/components/plots/styling/frequency/SelectCellSets.jsx:67]: Approved by Vicky 👍

File: src/components/plots/styling/frequency/SelectCellSets.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/plots/styling/frequency/SelectCellSets.jsx]: Done!

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx (hunk 6 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx]: Changed

File: src/__test__/components/plots/styling/frequency/SelectCellSets.test.jsx [hunk 9]
```diff
@@ -1,52 +1,91 @@
...
+      render(
+        <Provider store={storeState}>
+          {selectCellSetsFactory()}
+        </Provider>,
+      );
+    });
+
+    const dropdown = screen.getByRole('combobox', { name: 'groupBy' });
+
+    await act(async () => {
```
Comment (StefanBabukov) [inline src/__test__/components/plots/styling/frequency/SelectCellSets.test.jsx:79]: Do you need to wrap this in act? Did you get the warning for it?

File: src/pages/experiments/[experimentId]/plots-and-tables/frequency/index.jsx [hunk 10]
```diff
@@ -131,7 +120,12 @@ const frequencyPlot = ({ experimentId }) => {
 
     return (
       <center>
-        <FrequencyPlot hierarchy={hierarchy} properties={properties} config={config} />
+        <FrequencyPlot
+          experimentId={experimentId}
+          hierarchy={hierarchy}
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/frequency/index.jsx]: maybe you can just pass cellSets and destructure it inside the component? Less props this way.

File: src/redux/selectors.js [hunk 11]
```diff
@@ -17,13 +17,13 @@ const getCellSetsHierarchy = (...params) => (state) => (
...
-const getCellSetsHierarchyByKey = (...params) => (state) => (
-  cellSetsSelectors.getCellSetsHierarchyByKey(...params)(state.cellSets));
+const getCellSetsHierarchyByKeys = (...params) => (state) => (
+  cellSetsSelectors.getCellSetsHierarchyByKeys(...params)(state.cellSets));
 
 export {
   getBackendStatus,
   getCellSets,
   getCellSetsHierarchy,
   getCellSetsHierarchyByType,
```
Comment (StefanBabukov) [inline src/redux/selectors.js:27]: We should start grouping the selectors and export - `cellSetsSelectors` instead of defining every single one here. In the future this will become a big mess.

File: src/__test__/components/plots/styling/frequency/SelectCellSets.test.jsx (hunk 9 above)
Comment (StefanBabukov) [reply src/__test__/components/plots/styling/frequency/SelectCellSets.test.jsx:79]: Does it need to be an async function too?

Comment (aerlaut) [reply src/__test__/components/plots/styling/frequency/SelectCellSets.test.jsx:79]: Yes, I do get the warning if I don't wrap it in `act` or remove the `async`. I think the `async` is there because there are async functions that are dispatched after firing the event, and to handle it we need to make the callback `async` which is resolved by `await`.

File: src/pages/experiments/[experimentId]/plots-and-tables/frequency/index.jsx (hunk 10 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/frequency/index.jsx]: Good catch!

File: src/redux/selectors.js (hunk 11 above)
Comment (aerlaut) [reply src/redux/selectors.js:27]: This can already be done if we remove this file and import directly from `redux/selectors/{selectorGroup}`. I agree with you it's going to be messy in the long run. What do you think @cosa65?

Comment (aerlaut) [reply src/redux/selectors.js:27]: This can already be done if we remove this file and import it directly from `redux/selectors/{selectorGroup}`. I agree with you it's going to be messy in the long run. What do you think @cosa65

File: src/pages/experiments/[experimentId]/plots-and-tables/frequency/index.jsx (hunk 10 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/frequency/index.jsx]: I've changed it to use `cellSets`

**Review by vickymorrison — APPROVED** [review]

File: src/redux/selectors.js (hunk 11 above)
Comment (aerlaut) [reply src/redux/selectors.js:27]: I think this issue can be addressed in a different ticket.

**Review by StefanBabukov — APPROVED** [review]
lgtm
